client/my-sites/checkout/checkout/index.jsx

```diff
@@ -394,8 +396,34 @@ export class Checkout extends React.Component {
 		} = this.props;
 		const domainReceiptId = get( getGoogleApps( cart ), '[0].extra.receipt_for_domain', 0 );
 
+		const adminUrl = get( selectedSite, [ 'options', 'admin_url' ] );
+
+		// If we're given an explicit `redirectTo` query arg, make sure it's either internal
+		// (i.e. on WordPress.com), or a Jetpack or WP.com site's block editor (in wp-admin).
+		// This is required for Jetpack's (and WP.com's) paid blocks Upgrade Nudge.
 		if ( redirectTo ) {
-			return redirectTo;
+			if ( ! isExternal( redirectTo ) ) {
+				return redirectTo;
+			}
+
+			const { protocol, hostname, port, pathname, query } = parseUrl( redirectTo, true, true );
+
+			// We cannot simply compare `hostname` to `selectedSiteSlug`, since the latter
+			// might contain a path in the case of Jetpack subdirectory installs.
+			if ( redirectTo.startsWith( adminUrl ) && pathname.endsWith( '/wp-admin/post.php' ) ) {
```

Comment:
`adminUrl` should cover `/wp-admin/` part, and I'm pretty sure this is configurable so we shouldn't hardcode it.

Maybe a whitelist of known good admin routes?

```suggestion
			const allowedAdminRoutes =[
				'post.php',
				'post-new.php',
				// …
			];
			if ( adminUrl && allowedAdminRoutes.map( routePart => `${adminUrl}${routePart}` ).includes( redirectTo ) {
```

We might also need some `trailingSlashIt` magic there

Reply:
Thanks, makes sense :thinking: I just think I'll go without the whitelist for now (just using `post.php`) because YAGNI (and we don't need `post-new.php` b/c autosave).

Reply:
> I just think I'll go without the whitelist for now

Sure, that's fine. My comment had a few ideas, to be clear I think it's important to rework this check at least like:

```suggestion
			if ( adminUrl && `${adminUrl}post.php` ) === redirectTo ) {
```

That should help account for cases where admin url might not be `/wp-admin`

Reply:
Pushed f8444d0a81 which is essentially this ^ except it takes into account that `redirecTo` also has a query string.